src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -295,42 +295,95 @@ namespace winrt::Microsoft::Terminal::Control::implementation
+    winrt::fire_and_forget TermControl::UpdateAppearance(IControlAppearance newAppearance)
+    {
+        // Dispatch a call to the UI thread
+        co_await winrt::resume_foreground(Dispatcher());
+        _UpdateAppearanceFromUIThread(newAppearance);
+    }
+
+    // Method Description:
+    // - Writes the given sequence as input to the active terminal connection,
+    // Arguments:
+    // - wstr: the string of characters to write to the terminal connection.
+    // Return Value:
+    // - <none>
+    void TermControl::SendInput(const winrt::hstring& wstr)
+    {
+        _SendInputToConnection(wstr);
+    }
+
+    void TermControl::ToggleShaderEffects()
+    {
+        auto lock = _terminal->LockForWriting();
+        // Originally, this action could be used to enable the retro effects
+        // even when they're set to `false` in the settings. If the user didn't
+        // specify a custom pixel shader, manually enable the legacy retro
+        // effect first. This will ensure that a toggle off->on will still work,
+        // even if they currently have retro effect off.
+        if (_settings.PixelShaderPath().empty() && !_renderEngine->GetRetroTerminalEffect())
+        {
+            // SetRetroTerminalEffect to true will enable the effect. In this
+            // case, the shader effect will already be disabled (because neither
+            // a pixel shader nor the retro effects were originally requested).
+            // So we _don't_ want to toggle it again below, because that would
+            // toggle it back off.
+            _renderEngine->SetRetroTerminalEffect(true);
+        }
+        else
+        {
+            _renderEngine->ToggleShaderEffects();
+        }
+    }
+
+    // Method Description:
+    // - Updates the settings of the current terminal.
     // - This method is separate from UpdateSettings because there is an apparent optimizer
     //   issue that causes one of our hstring -> wstring_view conversions to result in garbage,
     //   but only from a coroutine context. See GH#8723.
     // - INVARIANT: This method must be called from the UI thread.
     // Arguments:
-    // - <none>
-    // Return Value:
-    // - <none>
-    void TermControl::_UpdateSettingsOnUIThread()
+    // - newSettings: the new settings to set
+    void TermControl::_UpdateSettingsFromUIThread(IControlSettings newSettings)
     {
         if (_closing)
         {
             return;
         }
 
-        auto lock = _terminal->LockForWriting();
-
         // Update our control settings
         _ApplyUISettings(_settings);
 
         // Update the terminal core with its new Core settings
         _terminal->UpdateSettings(_settings);
 
-        if (!_initializedTerminal)
-        {
-            // If we haven't initialized, there's no point in continuing.
-            // Initialization will handle the renderer settings.
-            return;
-        }
+        auto lock = _terminal->LockForWriting();

```

Comment:
Still mildly worried that this is below `_terminal->UpdateSettings(_settings);` - I'm pretty sure this should be above

Reply:
@PankajBhojwani yeah -- why did this lock move down?

Reply:
i am also very worried about the removal of the initialization check below. I didn't see it pop back up anywhere.

This lock guards access to the variable determining whether the terminal is initialized above all -- we *will* explode if we try to update settings while not initialized.

Reply:
Wait I thought I moved the lock back to the correct place when MG pointed it out. Fixed now, thanks for catching it!